packages/vanilla/src/task.ts

```diff
@@ -0,0 +1,245 @@
+	TaskFunction,
+} from '@sheepdog/core';
+import { CancelationError, createTask, handlers } from '@sheepdog/core';
+export type { SheepdogUtils, TaskOptions };
+
+export { CancelationError };
+
+export type Task<TArgs = unknown, TReturn = unknown> = ReturnType<typeof task<TArgs, TReturn>>;
+
+type InstanceEvents = 'error' | 'finish' | 'cancel' | 'start' | 'success';
+type Events = 'start' | 'finish' | 'instance-create' | `instance-${InstanceEvents}`;
+
+class SheepdogEvent extends CustomEvent<unknown> {
+	constructor(name: Events) {
+		super(name);
+	}
+}
+
+class SheepdogInstanceEvent extends CustomEvent<unknown> {
+	constructor(name: InstanceEvents) {
+		super(name);
+	}
+}
+
+type MappedInstance<TReturn> = {
+	event_target: EventTarget;
+	is_running: boolean;
+	instance: TaskInstance<TReturn>;
+	offs: Set<() => void>;
+};
+
+export type TaskInstance<TReturn = undefined> = {
+	error?: unknown;
+	value?: TReturn;
+	on: (event: InstanceEvents, cb: () => void, options?: AddEventListenerOptions) => void;
+};
+
+function define_properties<TOn, TFrom>(on: TOn, from: TFrom): asserts on is TOn & Readonly<TFrom> {
+	for (const key in from) {
+		Object.defineProperty(on, key, {
+			enumerable: true,
+			get() {
+				return from[key as never];
+			},
+		});
+	}
+}
+
+function _task<TArgs = unknown, TReturn = undefined>(
+	gen_or_fun: TaskFunction<TArgs, TReturn>,
+	options?: TaskOptions,
+) {
+	const event_target = new EventTarget();
+
+	const task_instance = {
+		isRunning: false,
+		last: undefined as undefined | TaskInstance<TReturn>,
+		lastCanceled: undefined as undefined | TaskInstance<TReturn>,
+		lastErrored: undefined as undefined | TaskInstance<TReturn>,
+		lastRunning: undefined as undefined | TaskInstance<TReturn>,
+		lastSuccessful: undefined as undefined | TaskInstance<TReturn>,
+		performCount: 0,
+		on(event: Events, cb: () => void, options?: AddEventListenerOptions) {
+			event_target.addEventListener(event, cb, options);
+			return () => {
+				event_target.removeEventListener(event, cb, options);
+			};
+		},
+		destroy() {},
+	};
+
+	function update_is_running() {
+		task_instance.isRunning = [...instances.values()].some((val) => val.is_running);
+	}
+
+	const instances = new Map<string, MappedInstance<TReturn>>();
+
+	const actual_task = createTask<TArgs, TReturn, TaskInstance<TReturn>>(
+		{
+			onDestroy(fn) {
+				task_instance.destroy = fn;
+			},
+			onError(instance_id, error) {
+				const instance = instances.get(instance_id);
+				if (instance) {
+					task_instance.lastErrored = instance.instance;
+					task_instance.lastRunning = undefined;
+					instance.instance.error = error;
+					instance.is_running = false;
+					update_is_running();
+					instance.event_target.dispatchEvent(new SheepdogInstanceEvent('error'));
+					instance.event_target.dispatchEvent(new SheepdogInstanceEvent('finish'));
+					// we delete after a microtask to avoid returnModifier
+					// not founding the instance in case of a syncronous
```

Comment:
```suggestion
					// not finding the instance in case of a synchronous
```

This is what I assume is supposed to be here 😬 please do check.

Reply:
It is lol